File: examples/tapis_secure/runner.py
# /// script
# requires-python = ">=3.9"
# dependencies = [
#   "requests-oauthlib>=2.0.0",
# ]
# ///

import argparse
import json
import os
import sys
from typing import Any

import requests
from oauthlib.oauth2 import BackendApplicationClient
from requests.auth import HTTPBasicAuth
from requests_oauthlib import OAuth2Session


CLIENT_ID = "goto-cli"
CLIENT_SECRET = "goto-secret"


def print_response(label: str, response: requests.Response) -> None:
    print()
    print(f"== {label}")
    print(f"status: {response.status_code}")
    content_type = response.headers.get("content-type")
    if content_type:
        print(f"content-type: {content_type}")
    www_authenticate = response.headers.get("www-authenticate")
    if www_authenticate:
        print(f"www-authenticate: {www_authenticate}")

    try:
        body: Any = response.json()
    except ValueError:
        print(f"body: {response.text}")
    else:
        print("body:", json.dumps(body, sort_keys=True))


def require_status(response: requests.Response, expected: int, label: str) -> None:
    if response.status_code != expected:
        print_response(label, response)
        raise SystemExit(f"{label} returned {response.status_code}, expected {expected}")


def oauth_client(base_url: str, scope: str) -> OAuth2Session:
    # The example server is intentionally local HTTP; production OAuth2 should use HTTPS.
    os.environ.setdefault("OAUTHLIB_INSECURE_TRANSPORT", "1")

    token_url = f"{base_url}/oauth/token"
    client = BackendApplicationClient(client_id=CLIENT_ID)
    session = OAuth2Session(client=client)
    token = session.fetch_token(
        token_url=token_url,
        auth=HTTPBasicAuth(CLIENT_ID, CLIENT_SECRET),
        scope=[scope],
    )
    print()
    print(f"== oauth2 client credentials token for {scope}")
    print("token_type:", token.get("token_type"))
    print("expires_in:", token.get("expires_in"))
    print("scope:", token.get("scope"))
    return session


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run a Python OAuth2 client against examples/tapis_secure/server.nim"
    )
    parser.add_argument("--base-url", default="http://127.0.0.1:9083")
    args = parser.parse_args()
    base_url = args.base_url.rstrip("/")

    print(f"Python OAuth2 client talking to {base_url}")

    health = requests.get(f"{base_url}/health", timeout=5)
    print_response("public health route", health)
    require_status(health, 200, "public health route")

    resolved = requests.get(
        f"{base_url}/go/docs", params={"preview": "true"}, timeout=5
    )
    print_response("public goto resolver", resolved)
    require_status(resolved, 200, "public goto resolver")

    unauthenticated = requests.get(f"{base_url}/admin/gotos", timeout=5)
    print_response("scoped route without bearer token", unauthenticated)
    require_status(unauthenticated, 401, "scoped route without bearer token")

    read_client = oauth_client(base_url, "goto:read")
    write_client = oauth_client(base_url, "goto:write")

    protected_list = read_client.get(
        f"{base_url}/admin/gotos", params={"limit": 2}, timeout=5
    )
    print_response("read-scoped admin list", protected_list)
    require_status(protected_list, 200, "read-scoped admin list")

    out_of_scope = read_client.get(
        f"{base_url}/admin/gotos/new/save",
        params={"url": "https://example.test/notes"},
        timeout=5,
    )
    print_response("write route with read-only token", out_of_scope)
    require_status(out_of_scope, 403, "write route with read-only token")

    saved = write_client.get(
        f"{base_url}/admin/gotos/new/save",
        params={"url": "https://example.test/notes", "title": "Notes"},
        timeout=5,
    )
    print_response("write-scoped save", saved)
    require_status(saved, 200, "write-scoped save")

    deleted = write_client.get(f"{base_url}/admin/gotos/new/delete", timeout=5)
    print_response("write-scoped delete", deleted)
    require_status(deleted, 200, "write-scoped delete")

    swagger = requests.get(f"{base_url}/swagger.json", timeout=5)
    require_status(swagger, 200, "openapi")
    spec = swagger.json()
    print()
    print("== openapi")
    print("title:", spec["info"]["title"])
    print("paths:", ", ".join(spec["paths"].keys()))
    print("security schemes:", ", ".join(spec["components"]["securitySchemes"].keys()))
    print(
        "admin list security:",
        json.dumps(spec["paths"]["/admin/gotos"]["get"]["security"]),
    )

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/tapis_secure/runner.nim
import std/[httpclient, json, os, osproc, sequtils, strformat, strutils]

proc repoRoot(): string =
  currentSourcePath.parentDir().parentDir().parentDir()

proc compileServer(repoDir: string, serverBinary: string) =
  let serverSource = repoDir / "examples" / "tapis_secure" / "server.nim"
  let compileResult = execCmdEx(
    "nim c -o:" & quoteShell(serverBinary) & " " & quoteShell(serverSource),
    options = {poUsePath, poStdErrToStdOut},
  )
  if compileResult.exitCode != 0:
    echo compileResult.output
    quit("failed to compile examples/tapis_secure/server.nim", QuitFailure)

proc waitUntilReady(baseUrl: string) =
  var client = newHttpClient(timeout = 500)
  defer:
    client.close()

  for attempt in 0 ..< 30:
    try:
      let response = client.get(baseUrl & "/health")
      if response.code.int == 200:
        return
    except CatchableError:
      discard
    sleep(200)

  quit("secure example server did not become ready", QuitFailure)

proc printResponse(label: string, response: Response) =
  echo ""
  echo "== ", label
  echo "status: ", response.code.int
  if response.headers.hasKey("Content-Type"):
    echo "content-type: ", response.headers["Content-Type"]
  if response.headers.hasKey("WWW-Authenticate"):
    echo "www-authenticate: ", response.headers["WWW-Authenticate"]
  echo "body: ", response.body

proc requestToken(client: HttpClient, baseUrl, scope: string): string =
  let response = client.request(
    baseUrl & "/oauth/token",
    httpMethod = HttpPost,
    headers = newHttpHeaders(
      {
        "Authorization": "Basic Z290by1jbGk6Z290by1zZWNyZXQ=",
        "Content-Type": "application/x-www-form-urlencoded",
      }
    ),
    body = "grant_type=client_credentials&scope=" & scope,
  )
  printResponse("oauth2 client credentials token for " & scope, response)
  if response.code.int != 200:
    quit("failed to issue token for scope " & scope, QuitFailure)

  parseJson(response.body)["access_token"].getStr()

proc authedGet(client: HttpClient, url, token: string): Response =
  client.request(
    url,
    httpMethod = HttpGet,
    headers = newHttpHeaders({"Authorization": "Bearer " & token}),
  )

proc main() =
  let repoDir = repoRoot()
  let serverBinary = repoDir / "examples" / "tapis_secure" / "server_bin"
  let port = 9083
  let baseUrl = fmt"http://127.0.0.1:{port}"

  compileServer(repoDir, serverBinary)
  defer:
    if fileExists(serverBinary):
      removeFile(serverBinary)

  let serverProc = startProcess(
    serverBinary,
    workingDir = repoDir,
    args = [$port],
    options = {poParentStreams, poStdErrToStdOut},
  )
  defer:
    if serverProc.running():
      terminate(serverProc)
      discard serverProc.waitForExit(3_000)
    serverProc.close()

  waitUntilReady(baseUrl)

  var client = newHttpClient(timeout = 5_000)
  defer:
    client.close()

  echo "TAPIS secure goto runner talking to ", baseUrl

  let health = client.get(baseUrl & "/health")
  printResponse("public health route", health)

  let resolved = client.get(baseUrl & "/go/docs?preview=true")
  printResponse("public goto resolver", resolved)

  let unauthenticated = client.get(baseUrl & "/admin/gotos")
  printResponse("scoped route without bearer token", unauthenticated)

  let readToken = requestToken(client, baseUrl, "goto%3Aread")
  let writeToken = requestToken(client, baseUrl, "goto%3Awrite")

  let protectedList = authedGet(client, baseUrl & "/admin/gotos?limit=2", readToken)
  printResponse("read-scoped admin list", protectedList)

  let outOfScope = authedGet(
    client,
    baseUrl & "/admin/gotos/new/save?url=https%3A%2F%2Fexample.test%2Fnotes",
    readToken,
  )
  printResponse("write route with read-only token", outOfScope)

  let saved = authedGet(
    client,
    baseUrl &
      "/admin/gotos/new/save?url=https%3A%2F%2Fexample.test%2Fnotes&title=Notes",
    writeToken,
  )
  printResponse("write-scoped save", saved)

  let deleted = authedGet(client, baseUrl & "/admin/gotos/new/delete", writeToken)
  printResponse("write-scoped delete", deleted)

  let broken = client.get(baseUrl & "/broken")
  printResponse("exception converted to tapis error", broken)

  let swagger = client.get(baseUrl & "/swagger.json")
  let spec = parseJson(swagger.body)
  echo ""
  echo "== openapi"
  echo "title: ", spec["info"]["title"].getStr()
  echo "paths: ", spec["paths"].keys().toSeq().join(", ")
  echo "security schemes: ",
    spec["components"]["securitySchemes"].keys().toSeq().join(", ")
  echo "admin list security: ", $spec["paths"]["/admin/gotos"]["get"]["security"]

when isMainModule:
  main()
